PolyEngine/Core/Src/String.cpp

```diff
@@ -0,0 +1,207 @@
+		if (c > 64 && c < 91) {
+			c += 32;
+		}
+		d.PushBack(c);
+	}
+	return new String(d.GetData());
+}
+
+String* String::ToUpper() const {
+	Dynarray<char> d = Dynarray<char>(this->Data.GetSize());
+	for (int i = 0; i < this->Data.GetSize(); i++) {
+		char c = this->Data[i];
+		if (c > 96 && c < 123) {
+			c -= 32;
+		}
+		d.PushBack(c);
+	}
+	return new String(d.GetData());
+}
+
+bool String::IsEmpty() const {
+	return (bool)GetLength();
+}
+
+String* String::Replace(char a, char b) const {
+	Dynarray<char> d = Dynarray<char>(this->Data.GetSize());
+	for (int i = 0; i < this->Data.GetSize(); i++) {
+		char c = this->Data[i];
+		if (c == a) {
+			c = b;
+		}
+		d.PushBack(c);
+	}
+	return new String(d.GetData());
+}
+
+String* String::Replace(const String& a, const String& b) const {
+	//TODO Dynarray.Contains(Dynarray<T> rhs)
+	return nullptr;
+}
+
+String* String::Split(char delimiter) const {
+	//TODO Dynarray.FindAllIdx(T rhs)
+	return nullptr;
+}
+
+String* String::Join(const String& vars, size_t size, const String& separator) {
+	//TODO stringbuilder
+	return nullptr;
+}
+
+bool String::StartsWith(char var) const {
+	return (Data[0] == var);
+}
+
+bool String::EndsWith(char var) const {
+	return (Data[Data.GetSize()-2] == var);
+}
+
+String* String::Substring(int end) const {
+	return Substring(0, end);
+}
+
+String* String::Substring(int start, int end) const {
+	//TODO dynarray constructor?
+	Dynarray<char> d = Dynarray<char>(end - start + 1);
+	for (int i = start; i < end; i++) {
+		d.PushBack(this->Data[i]);
+	}
+	d.PushBack(0); //why is this needed? constructor should do that
+	return new String(d.GetData());
+}
+
+
+String* String::GetTrimmed() const {
+	//TODO substrings
+	Dynarray<char> d = this->Data;
+	while (d[0] == ' ' || d[0] == '\n' || d[0] == '\r' || d[0] == '\t') { //convert whitespaces to String and use Contains(d[0])?
+		d.PopFront();
+	}
+	d.PopBack(); //pop 0
+	while (d[d.GetSize()-1] == ' ' || d[d.GetSize()-1] == '\n' || d[d.GetSize()-1] == '\r' || d[d.GetSize()-1] == '\t') {
+		d.PopBack();
+	}
+	d.PushBack(0);
+
+	return new String(d.GetData());
+}
+
+
+String& String::operator=(const String& rhs) {
+	Data = rhs.Data;
+	return *this;
+}
+
+String& String::operator=(String&& rhs) {
+	Data = std::move(rhs.Data);
+	return *this;
+}
+
+bool String::operator==(const char* str) {
+	if (GetLength() != StrLen(str))
+		return false;
+	for (size_t k = 0; k < GetLength(); ++k)
+		if (Data[k] != str[k])
+			return false;
+	return true;
+}
+
+bool String::operator==(const String& str) {
+	return Data == str.Data;
+}
+
+bool String::operator<(const String& rhs) const {
+	if (GetLength() < rhs.GetLength())
+		return true;
+	else if (GetLength() > rhs.GetLength())
+		return false;
+
+	for (size_t i = 0; i < GetLength(); ++i)
+	{
+		if (Data[i] < rhs.Data[i])
+			return true;
+		else if (Data[i] > rhs.Data[i])
+			return false;
+	}
+	return false;
+}
+
+String String::operator+(const String& rhs) const {
+	String ret;
```

Comment:
Here you have an example of how to remove double allocation that was present in previous code.